apm-agent-core/src/main/java/co/elastic/apm/agent/util/CallDepth.java

```diff
@@ -0,0 +1,134 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2020 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.util;
+
+import net.bytebuddy.asm.Advice;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * A utility that makes it easy to detect nested method calls.
+ */
+public class CallDepth {
+    private static final ConcurrentMap<String, CallDepth> registry = new ConcurrentHashMap<>();
+    private final ThreadLocal<MutableInt> callDepthPerThread = new ThreadLocal<MutableInt>();
```

Comment:
Unless you are expecting a nesting of over 127 levels of the same advice, you should be able to rely on the Integer cache in order to avoid allocations related to boxing.

Reply:
Both seem like valid approaches with minor differences in performance and lines of code.

Reply:
I ended up implementing your suggestion in https://github.com/elastic/apm-agent-java/pull/1273 as it may introduce thread local class loader leaks down the road.